### PR TITLE
samples: fix flaky tests

### LINKED ISSUE #227: samples.snippets.subscriber_test: test_remove_dead_letter_policy failed
This test failed!

To configure my behavior, see [the Build Cop Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/master/packages/buildcop).

If I'm commenting on this issue too often, add the `buildcop: quiet` label and
I will stop commenting.

---

commit: 7de13b834cb4d015730fcee711711d13e7a28112
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/042e0a2a-44d0-470c-a7ce-6118a3aae390), [Sponge](http://sponge2/042e0a2a-44d0-470c-a7ce-6118a3aae390)
status: failed
<details><summary>Test output</summary><br><pre>args = (subscription: "projects/python-docs-samples-tests/subscriptions/subscription-test-subscription-dlq-94797d067923478cb47afe733b09d01a"
,)
kwargs = {'metadata': [('x-goog-request-params', 'subscription=projects/python-docs-samples-tests/subscriptions/subscription-test-subscription-dlq-94797d067923478cb47afe733b09d01a'), ('x-goog-api-client', 'gl-python/3.8.3 grpc/1.32.0 gax/1.23.0')]}

    @six.wraps(callable_)
    def error_remapped_callable(*args, **kwargs):
        try:
>           return callable_(*args, **kwargs)

.nox/py-3-8/lib/python3.8/site-packages/google/api_core/grpc_helpers.py:57: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <grpc._channel._UnaryUnaryMultiCallable object at 0x7f24afc6d6a0>
request = subscription: "projects/python-docs-samples-tests/subscriptions/subscription-test-subscription-dlq-94797d067923478cb47afe733b09d01a"

timeout = None
metadata = [('x-goog-request-params', 'subscription=projects/python-docs-samples-tests/subscriptions/subscription-test-subscription-dlq-94797d067923478cb47afe733b09d01a'), ('x-goog-api-client', 'gl-python/3.8.3 grpc/1.32.0 gax/1.23.0')]
credentials = None, wait_for_ready = None, compression = None

    def __call__(self,
                 request,
                 timeout=None,
                 metadata=None,
                 credentials=None,
                 wait_for_ready=None,
                 compression=None):
        state, call, = self._blocking(request, timeout, metadata, credentials,
                                      wait_for_ready, compression)
>       return _end_unary_response_blocking(state, call, False, None)

.nox/py-3-8/lib/python3.8/site-packages/grpc/_channel.py:826: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

state = <grpc._channel._RPCState object at 0x7f24acb35a00>
call = <grpc._cython.cygrpc.SegregatedCall object at 0x7f24afbb06c0>
with_call = False, deadline = None

    def _end_unary_response_blocking(state, call, with_call, deadline):
        if state.code is grpc.StatusCode.OK:
            if with_call:
                rendezvous = _MultiThreadedRendezvous(state, call, None, deadline)
                return state.response, rendezvous
            else:
                return state.response
        else:
>           raise _InactiveRpcError(state)
E           grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
E           	status = StatusCode.UNKNOWN
E           	details = "Stream removed"
E           	debug_error_string = "{"created":"@1603278725.494860777","description":"Error received from peer ipv4:74.125.195.95:443","file":"src/core/lib/surface/call.cc","file_line":1061,"grpc_message":"Stream removed","grpc_status":2}"
E           >

.nox/py-3-8/lib/python3.8/site-packages/grpc/_channel.py:729: _InactiveRpcError

The above exception was the direct cause of the following exception:

subscriber_client = <google.cloud.pubsub_v1.SubscriberClient object at 0x7f24afc81b50>
topic = 'projects/python-docs-samples-tests/topics/subscription-test-topic-94797d067923478cb47afe733b09d01a'

    @pytest.fixture(scope="module")
    def subscription_dlq(subscriber_client, topic):
        subscription_path = subscriber_client.subscription_path(PROJECT, SUBSCRIPTION_DLQ)
    
        try:
            subscription = subscriber_client.get_subscription(
                request={"subscription": subscription_path}
            )
        except:  # noqa
            subscription = subscriber_client.create_subscription(
                request={"name": subscription_path, "topic": topic}
            )
    
        yield subscription.name
    
>       subscriber_client.delete_subscription(request={"subscription": subscription.name})

subscriber_test.py:144: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../google/cloud/pubsub_v1/_gapic.py:40: in <lambda>
    fx = lambda self, *a, **kw: wrapped_fx(self.api, *a, **kw)  # noqa
../../google/pubsub_v1/services/subscriber/client.py:723: in delete_subscription
    rpc(
.nox/py-3-8/lib/python3.8/site-packages/google/api_core/gapic_v1/method.py:145: in __call__
    return wrapped_func(*args, **kwargs)
.nox/py-3-8/lib/python3.8/site-packages/google/api_core/retry.py:281: in retry_wrapped_func
    return retry_target(
.nox/py-3-8/lib/python3.8/site-packages/google/api_core/retry.py:184: in retry_target
    return target()
.nox/py-3-8/lib/python3.8/site-packages/google/api_core/grpc_helpers.py:59: in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

value = None
from_value = <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.UNKNOWN
	details = "Stream removed"
	debug_error_....195.95:443","file":"src/core/lib/surface/call.cc","file_line":1061,"grpc_message":"Stream removed","grpc_status":2}"
>

>   ???
E   google.api_core.exceptions.Unknown: None Stream removed

<string>:3: Unknown</pre></details>

### PR DESCRIPTION
**Pull request by anguillanneuf**
The two tests at the end of `subscriber_tests.py` often cause tests to fail due to subscription `NotFound` error. 

The order of the tests shouldn't matter, what really fixes the tests is the python version in resource names. I suspect the different runtimes were pulling messages from the same topic (created using UUID) and destroying the topic and subscription resources before the other runtimes could finish. Or the service takes too long to make topics and subscriptions (sometimes up to two minutes).

Fixes #227 #240 #246

### CODE REVIEW

**Comment by leahecole** [comment]
I know this is a draft but I'm just commenting because I'm mildly concerned about the title - tests should be order independent. If something needs to exist in order for a test to run, please use pytest fixtures to do so. Or, if it's something really long running (i.e. automl trained model) we may need to explore having a static resource.

And if you already knew all of this, feel free to ignore me. and if you want to talk more ping me :)

File: samples/snippets/subscriber_test.py [hunk 1]
```diff
@@ -161,6 +163,13 @@ def subscription_dlq(subscriber_client, topic, dead_letter_topic):
     subscriber_client.delete_subscription(request={"subscription": subscription.name})
 
 
+def _publish_messages(publisher_client, topic, **attrs):
+    for n in range(5):
+        data = "message {}".format(n).encode("utf-8")
```
Comment (leahecole) [inline samples/snippets/subscriber_test.py:168]:
```suggestion
        data = f"message {n}".encode("utf-8")
```

File: samples/snippets/subscriber_test.py [hunk 2]
```diff
@@ -23,13 +24,14 @@
...
-SUBSCRIPTION_ADMIN = "subscription-test-subscription-admin-" + UUID
-SUBSCRIPTION_ASYNC = "subscription-test-subscription-async-" + UUID
-SUBSCRIPTION_SYNC = "subscription-test-subscription-sync-" + UUID
-SUBSCRIPTION_DLQ = "subscription-test-subscription-dlq-" + UUID
+TOPIC = "subscription-test-topic-" + UUID + PY_VERSION
+DEAD_LETTER_TOPIC = "subscription-test-dead-letter-topic-" + PY_VERSION + "-" + UUID
+SUBSCRIPTION_ADMIN = "subscription-test-subscription-admin-" + PY_VERSION + "-" + UUID
+SUBSCRIPTION_ASYNC = "subscription-test-subscription-async-" + PY_VERSION + "-" + UUID
+SUBSCRIPTION_SYNC = "subscription-test-subscription-sync-" + PY_VERSION + "-" + UUID
+SUBSCRIPTION_DLQ = "subscription-test-subscription-dlq-" + PY_VERSION + "-" + UUID
```
Comment (leahecole) [inline samples/snippets/subscriber_test.py:34]: any reason in particular not to use f strings here too?

**Review by leahecole — APPROVED** [review]
Super minor nits, bu otherwise lgtm

File: samples/snippets/subscriber_test.py [hunk 3]
```diff
@@ -23,13 +24,14 @@
...
 UUID = uuid.uuid4().hex
+PY_VERSION = f"{sys.version_info.major}.{sys.version_info.minor}"
 PROJECT_ID = os.environ["GOOGLE_CLOUD_PROJECT"]
-TOPIC = "subscription-test-topic-" + UUID
-DEAD_LETTER_TOPIC = "subscription-test-dead-letter-topic-" + UUID
-SUBSCRIPTION_ADMIN = "subscription-test-subscription-admin-" + UUID
-SUBSCRIPTION_ASYNC = "subscription-test-subscription-async-" + UUID
-SUBSCRIPTION_SYNC = "subscription-test-subscription-sync-" + UUID
-SUBSCRIPTION_DLQ = "subscription-test-subscription-dlq-" + UUID
+TOPIC = "subscription-test-topic-" + UUID + PY_VERSION
```
Comment (pradn) [inline samples/snippets/subscriber_test.py]: Probably unnecessary to include PY_VERSION if there's already a UUID.

File: samples/snippets/subscriber.py [hunk 4]
```diff
@@ -540,9 +545,10 @@ def synchronous_pull_with_lease_management(project_id, subscription_id):
     subscriber = pubsub_v1.SubscriberClient()
     subscription_path = subscriber.subscription_path(project_id, subscription_id)
 
-    response = subscriber.pull(
-        request={"subscription": subscription_path, "max_messages": 3}
-    )
+    while not response:
```
Comment (pradn) [inline samples/snippets/subscriber.py:548]: Perhaps have some limit so this doesn't take forever? Maybe use time, and just try for 5 min max?

**Review by pradn — CHANGES_REQUESTED** [review]

File: samples/snippets/subscriber.py (hunk 4 above)
Comment (anguillanneuf) [reply samples/snippets/subscriber.py]: I set the retry deadline instead of using a while loop.

File: samples/snippets/subscriber_test.py (hunk 3 above)
Comment (anguillanneuf) [reply samples/snippets/subscriber_test.py]: I agree. Let me remove these. I pray for tests still pass.

Comment (anguillanneuf) [reply samples/snippets/subscriber_test.py]: No, they failed randomly for one of the runtimes. I'll leave them in.

**Review by pradn — APPROVED** [review]

**Comment by anguillanneuf** [comment]
Fixes #246

**Pull request merged**